tar.go

```diff
@@ -210,25 +234,43 @@ func untar(tr *tar.Reader, destination string) error {
 			return err
 		}
 
-		if err := untarFile(tr, header, destination); err != nil {
+		if err := untarFile(tr, header, destination, preserveIds); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
 // untarFile untars a single file from tr with header header into destination.
-func untarFile(tr *tar.Reader, header *tar.Header, destination string) error {
+func untarFile(tr *tar.Reader, header *tar.Header, destination string, preserveIds bool) error {
+	var err error
 	switch header.Typeflag {
 	case tar.TypeDir:
-		return mkdir(filepath.Join(destination, header.Name))
+		err = mkdir(filepath.Join(destination, header.Name))
+		break
 	case tar.TypeReg, tar.TypeRegA, tar.TypeChar, tar.TypeBlock, tar.TypeFifo:
-		return writeNewFile(filepath.Join(destination, header.Name), tr, header.FileInfo().Mode())
+		err = writeNewFile(filepath.Join(destination, header.Name), tr, header.FileInfo().Mode())
+		break
 	case tar.TypeSymlink:
-		return writeNewSymbolicLink(filepath.Join(destination, header.Name), header.Linkname)
+		err = writeNewSymbolicLink(filepath.Join(destination, header.Name), header.Linkname)
+		break
 	case tar.TypeLink:
-		return writeNewHardLink(filepath.Join(destination, header.Name), filepath.Join(destination, header.Linkname))
+		err = writeNewHardLink(filepath.Join(destination, header.Name), filepath.Join(destination, header.Linkname))
+		break
 	default:
-		return fmt.Errorf("%s: unknown type flag: %c", header.Name, header.Typeflag)
+		err = fmt.Errorf("%s: unknown type flag: %c", header.Name, header.Typeflag)
+		break
+	}
+	if err != nil {
+		return err
+	}
+
+	if preserveIds && header.Name != "./" && header.Name != "." && runtime.GOOS != "windows" {
+		if err = os.Lchown(filepath.Join(destination, header.Name), header.Uid, header.Gid); err != nil {
+			return err
+		}
+		// If a special permission is set usage of the syscall is required.
+		return syscall.Chmod(filepath.Join(destination, header.Name), uint32(header.Mode))
```

Comment:
One of the problems that invoking syscall.Chmod() and friends has, is that they indiscriminately will work on what you tell them to.  I can see two possible issues with the above:
  - if a previous entry in the archive was a softlink with a target to something outside the archive open tree, this later chmod(2) would end up changing the referenced file
  - the header.Name may be malicious, and point outside the destination tree, which if root uses this, could cause issues

I think for this PR to make headway, we would need to be much more careful with our paths, the cleaning of them, how we handle them, and have appropriate testing to make sure we're not going to end up with issues.